### PR TITLE
Add support for calculating convolution parameters when using atrous/dilated convolution

### PR DESCRIPTION
**Pull request by oveddan**
This PR will be the start of adding support for 2d Atrous convolution, as described in [Rethinking Atrous Convolution for Semantic Image Segmentation](https://arxiv.org/abs/1706.05587), by modifying `conv_util.computeConv2DInfo` to take arguments for dilation height and width and calculate the convolution info when there is dilation/atrous convolution.

Atrous convolution is currently available in Tensorflow, either by using [tf.nn.conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d) and setting the `strides` parameter, or by using [tf.nn.atrous_conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/atrous_conv2d) which is a wrapper around `conv2d`.

---
This change is [Reviewable](https://reviewable.io/reviews/pair-code/deeplearnjs/767)

### CODE REVIEW

**Comment by oveddan** [comment]
I signed it

**Comment by nsthorat** [comment]
It looks like depthwise conv2d and conv2d in TF have "rates" and "dilation" respectively for the arguments. I'm fine with choosing one internally for the ConvInfo interface, but when we do make the API change to add them to the signature of those methods, let's name them the same way TF does.

This LGTM overall though, minor nits :)

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/math.ts, line 371 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767#-L627_dTDgm0rBy4umCP:-L627_dTDgm0rBy4umCQ:b-h72a56) ([raw file](https://github.com/pair-code/deeplearnjs/blob/b79fb9aa7e05d4584f3b99b404802e2b7cc9dd66/src/math.ts#L371)):*
> ```XML
>               `${bias.rank}.`);
>     }
>     const res = ops.conv2d(x, filter, strides, 1, pad, dimRoundingMode);
> ```

where you simply pass a "1" as a literal, can you move it to a variable for readability?

`
const dilation = 1
const res = ops.conv2d(x, filter, strides, dilation, pad, dimRoundingMode);
`

---

*[src/ops/conv.ts, line 122 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767#-L5vSsz0xpdJD930mZVq:-L5vSsz0xpdJD930mZVr:b-fytksl) ([raw file](https://github.com/pair-code/deeplearnjs/blob/b79fb9aa7e05d4584f3b99b404802e2b7cc9dd66/src/ops/conv.ts#L122)):*
> ```XML
>   static conv2d<T extends Tensor3D|Tensor4D>(
>       x: T, filter: Tensor4D, strides: [number, number]|number,
>       rates: [number, number]|number, pad: 'valid'|'same'|number,
> ```

Can we hold off on this API change for now, and just keep working on internals until that's done and then expose it?

Same as in tensor.ts

You could below just set rates = 1 before passing to computeConv2dInfo.

---

*[src/ops/conv_util.ts, line 199 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767#-L627S5fgk36Bk6m3Gc4:-L627S5fgk36Bk6m3Gc5:b-clkk7g) ([raw file](https://github.com/pair-code/deeplearnjs/blob/b79fb9aa7e05d4584f3b99b404802e2b7cc9dd66/src/ops/conv_util.ts#L199)):*
> ```XML
>  */
> function getEffectiveFilterSize(filterSize: number, rate: number) {
>   if (rate <= 1) return filterSize;
> ```

use {} (part of our style guide)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767#-:-L627qTpMayVuq-NXKen:b-714ks1)*

**Comment by nsthorat** [comment]
:lgtm_strong:

Wil merge after these two small things. Thanks for the PR, sorry for the delay!

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/tensor.ts, line 860 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767#-L67L-6rbVSsOL-XJ8GS:-L67L-6rbVSsOL-XJ8GT:b-ihfewq) ([raw file](https://github.com/pair-code/deeplearnjs/blob/2823a7159a647b0d481ece515fa2fde6db25a762/src/tensor.ts#L860)):*
> ```XML
>   depthwiseConv2D<T extends Tensor3D|Tensor4D>(
>       this: T, filter: Tensor4D, strides: [number, number]|number,
>       pad: 'valid'|'same'|number, dilations: [number, number]|number = [1, 1],
> ```

sorry to be pedantic - can we keep this rates (just so we mirror TF exactly)?

---

*[src/ops/conv_util.ts, line 102 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767#-L67LRMKxRilKhFyvfXy:-L67LRMKxRilKhFyvfXz:b-4pa4r1) ([raw file](https://github.com/pair-code/deeplearnjs/blob/2823a7159a647b0d481ece515fa2fde6db25a762/src/ops/conv_util.ts#L102)):*
> ```XML
>     = getEffectiveFilterSize(filterWidth, dilationWidth);
> 
>   const { padInfo, outHeight, outWidth } = getPadAndOutInfo(
> ```

can you make sure you're using the right autoformatter (if you're using one)? If using VSCode, make sure you have tslint extension and clang-format extension, along with the clang-format command line tool.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767#-:-L67M5ozzAlpICglYUK0:b-uw54km)*

**Comment by oveddan** [comment]
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/tensor.ts, line 860 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767#-L67L-6rbVSsOL-XJ8GS:-L67mMN_uV4FZ9EfxiLT:btcuiz3) ([raw file](https://github.com/pair-code/deeplearnjs/blob/2823a7159a647b0d481ece515fa2fde6db25a762/src/tensor.ts#L860)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

sorry to be pedantic - can we keep this rates (just so we mirror TF exactly)?
</blockquote></details>

Totally see the need to have the api here mirror TF's.  However, in TF, it is named inconsistently, in some places `rate(s)` and others `dilation(s)`:

* [tf.nn.conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d) - `dilations`
* [tf.nn.conv2d_backprop_filter](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d_backprop_filter) - `dilations`
* [tf.nn.conv2d_backprop_input](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d_backprop_input) - `dilations`
* [tf.nn.conv2d_transpose](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d_transpose) - not a parameter
* [tf.nn.conv3d](https://www.tensorflow.org/api_docs/python/tf/nn/conv3d) - `dilations`
* [tf.nn.conv3d_backprop_filter_v2](https://www.tensorflow.org/api_docs/python/tf/nn/conv3d_backprop_filter_v2) - `dilations`
* [tf.nn.convolution](https://www.tensorflow.org/api_docs/python/tf/nn/convolution) - `dilation_rate`
* [tf.nn.conv1d](https://www.tensorflow.org/api_docs/python/tf/nn/conv1d) - not a parameter
* [tf.nn.atrous_conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/atrous_conv2d) -`rate`
* [tf.nn.atrous_conv2d_transpose](https://www.tensorflow.org/api_docs/python/tf/nn/atrous_conv2d_transpose) - `rate`
* [tf.nn.depthwise_conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d) - `rate`
* [tf.nn.depthwise_conv2d_native](https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d_native) - `dilations`
* [tf.nn.depthwise_conv2d_native_backprop_filter](https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d_native_backprop_filter) - `dilations`
* [tf.nn.depthwise_conv2d_native_backprop_input](https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d_native_backprop_input) - `dilations`
* [tf.nn.dilation2d](https://www.tensorflow.org/api_docs/python/tf/nn/dilation2d) - `rates`
* [tf.nn.quantized_conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/quantized_conv2d) - `dilations`

I just bring up the above list to highlight the inconsistencies in TF.   I feel it would make the deeplearn.js methods clearer to understand if the naming of that parameter was consistent within deeplearn. 

Also, in TF, [tf.nn.depthwise_conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d) only allows a single value for dilation rate, but it wouldn't be challenging to implement it on our end allowing for both width and height to be specified in the dilation rate.  If we are to keep consistent with TF that argument in `depthwiseConv2d` should be changed to just accept a single value.

Let me know your thoughts.

---

*[src/ops/conv_util.ts, line 102 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767#-L67LRMKxRilKhFyvfXy:-L67lqPpjAmnp0jPPYrU:b-19zw7a) ([raw file](https://github.com/pair-code/deeplearnjs/blob/2823a7159a647b0d481ece515fa2fde6db25a762/src/ops/conv_util.ts#L102)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

can you make sure you're using the right autoformatter (if you're using one)? If using VSCode, make sure you have tslint extension and clang-format extension, along with the clang-format command line tool.
</blockquote></details>

fixed 40c047ffafe995f0c5023c6303b377a0d41f2a5a.
I have been using VSCode but disabled the autoformatter because it was causing some weird behavior, like 
removing spaces in lines that I had not touched.  But now with it enabled, along with clang-format it works well.  Ive gone ahead and ran it on all the files that were changed as part of this commit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767)*

**Comment by oveddan** [comment]
Let me know what you think regarding the consistency with TF Api.  Also - should I squash all the commits into one?

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767#-:-L67qNqwLWaWOkCVCOTF:b8bzqa6)*

File: src/ops/conv.ts [hunk 1]
```diff
@@ -365,11 +373,11 @@ export class ConvOps {
...
-   * @param rates The dilation rates: `[rateHeight, rateWidth]` in which we
-   *     sample input values across the height and width dimensions in atrous
-   *     convolution. Defaults to `[1, 1]`. If `rate` is a single number, then
-   *     `rateHeight == rateWidth`. If it is greater than 1, then all values
-   *     of `strides` must be 1.
+   * @param dilations The dilation rates: `[dilationHeight, dilationWidth]`
+   *     in which we sample input values across the height and width dimensions
+   *     in atrous convolution. Defaults to `[1, 1]`. If `dilations` is a single
+   *     number, then `dilationHeight == dilationWidth`. If it is greater than
+   *     1, then all values of `strides` must be 1.
```
Comment (gpapan) [inline src/ops/conv.ts]: The constraint "if dilations > 1 then strides must be equal to 1" is a limitation of the older indirect "space_to_batch -> conv -> batch_to_space" implementation approach in TF. This constraint should not apply and can be lifted in your current implementation, although typically the constraint is valid in most use cases.

Comment (oveddan) [reply src/ops/conv.ts]: @gpapan 👍

Comment (oveddan) [reply src/ops/conv.ts]: @gpapan the description of this param was there prior to this PR.   All that changed here was the param name.  I will fix it with the other PR that implements atrous convolution.

Comment (gpapan) [reply src/ops/conv.ts]: Sounds good!

**Comment by oveddan** [comment]
@nsthorat Just squashed the commits and pushed.  Still yet to be resolved is the naming of rate/dilation.

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767#-:-L6NVhBKFJTdNSbnuWiJ:b-lucuro)*

**Comment by dsmilkov** [comment]
:lgtm_strong:

---

Reviewed 1 of 6 files at r3, 1 of 5 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/tensor.ts, line 860 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767#-L67L-6rbVSsOL-XJ8GS:-L6Ru7cLoclkij7JJysd:ba59ubw) ([raw file](https://github.com/pair-code/deeplearnjs/blob/2823a7159a647b0d481ece515fa2fde6db25a762/src/tensor.ts#L860)):*
<details><summary><i>Previously, oveddan (Dan Oved) wrote…</i></summary><blockquote>

Totally see the need to have the api here mirror TF's.  However, in TF, it is named inconsistently, in some places `rate(s)` and others `dilation(s)`:

* [tf.nn.conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d) - `dilations`
* [tf.nn.conv2d_backprop_filter](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d_backprop_filter) - `dilations`
* [tf.nn.conv2d_backprop_input](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d_backprop_input) - `dilations`
* [tf.nn.conv2d_transpose](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d_transpose) - not a parameter
* [tf.nn.conv3d](https://www.tensorflow.org/api_docs/python/tf/nn/conv3d) - `dilations`
* [tf.nn.conv3d_backprop_filter_v2](https://www.tensorflow.org/api_docs/python/tf/nn/conv3d_backprop_filter_v2) - `dilations`
* [tf.nn.convolution](https://www.tensorflow.org/api_docs/python/tf/nn/convolution) - `dilation_rate`
* [tf.nn.conv1d](https://www.tensorflow.org/api_docs/python/tf/nn/conv1d) - not a parameter
* [tf.nn.atrous_conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/atrous_conv2d) -`rate`
* [tf.nn.atrous_conv2d_transpose](https://www.tensorflow.org/api_docs/python/tf/nn/atrous_conv2d_transpose) - `rate`
* [tf.nn.depthwise_conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d) - `rate`
* [tf.nn.depthwise_conv2d_native](https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d_native) - `dilations`
* [tf.nn.depthwise_conv2d_native_backprop_filter](https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d_native_backprop_filter) - `dilations`
* [tf.nn.depthwise_conv2d_native_backprop_input](https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d_native_backprop_input) - `dilations`
* [tf.nn.dilation2d](https://www.tensorflow.org/api_docs/python/tf/nn/dilation2d) - `rates`
* [tf.nn.quantized_conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/quantized_conv2d) - `dilations`

I just bring up the above list to highlight the inconsistencies in TF.   I feel it would make the deeplearn.js methods clearer to understand if the naming of that parameter was consistent within deeplearn. 

Also, in TF, [tf.nn.depthwise_conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d) only allows a single value for dilation rate, but it wouldn't be challenging to implement it on our end allowing for both width and height to be specified in the dilation rate.  If we are to keep consistent with TF that argument in `depthwiseConv2d` should be changed to just accept a single value.

Let me know your thoughts.
</blockquote></details>

`dilations` sounds good to me

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/767#-:-L6Ruuxjj2UP5SBDUBkm:b-jbsf8z)*

**Pull request merged**